src/PolyfillTrait6.php

```diff
@@ -0,0 +1,83 @@
+<?php
+
+/*
+ * This file is part of PHPUnit Good Practices.
+ *
+ * (c) Dariusz Rumiński <dariusz.ruminski@gmail.com>
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace PHPUnitGoodPractices\Polyfill;
+
+/**
+ * @internal
+ */
+trait PolyfillTrait6
+{
+    public function expectException($exception)
+    {
+        if (\is_callable(['parent', 'expectException'])) {
+            parent::expectException($exception);
+        } else {
+            $this->setExpectedException($exception);
+        }
+    }
+
+    public function expectExceptionMessageMatches($regexp)
+    {
+        if (\is_callable(['parent', 'expectExceptionMessageMatches'])) {
+            parent::expectExceptionMessageMatches($regexp);
+        } elseif (\is_callable(['parent', 'setExpectedExceptionRegExp'])) {
+            $this->setExpectedExceptionRegExp(\Exception::class, $regexp);
```

Comment:
`setExpectedExceptionRegExp` was removed in v6
let's try to use `expectExceptionMessageRegExp` first, and only if not available fallback to `setExpectedExceptionRegExp`

idea: we try different methods in order what is the newest approach

Reply:
This is indeed looks funny. But PHPUnit 5.7.27 won't catch an exception if you use _only_ `expectExceptionMessageRegExp`, which it has. It's actually pretty hilarious.

https://github.com/sebastianbergmann/phpunit/blob/b7803aeca3ccb99ad0a506fa80b64cd6a56bbc0c/src/Framework/TestCase.php#L550-L562

https://github.com/sebastianbergmann/phpunit/blob/b7803aeca3ccb99ad0a506fa80b64cd6a56bbc0c/src/Framework/TestCase.php#L617-L624

Reply:
For the record, I like the same idea, but it just doesn't work here.

Reply:
`expectExceptionMessageMatches` is setting the message only
`expectExceptionMessageRegExp` works exactly same way

`setExpectedExceptionRegExp` is also setting the exception class, not only exception message


you say that `expectExceptionMessageRegExp`, when used alone, won't set an exception class in first place. it is exactly how new method, `expectExceptionMessageMatches`, works

i believe we should `setExpectedExceptionRegExp` as last resort

Reply:
It is not. Compare their implementations:

https://github.com/sebastianbergmann/phpunit/blob/b7803aeca3ccb99ad0a506fa80b64cd6a56bbc0c/src/Framework/TestCase.php#L550-L562

https://github.com/sebastianbergmann/phpunit/blob/b7803aeca3ccb99ad0a506fa80b64cd6a56bbc0c/src/Framework/TestCase.php#L617-L624 

I did initially like you proposing, but it doesn't work. Just look at the code!

Reply:
Look:

https://github.com/PHPUnitGoodPractices/polyfill/pull/17/commits/5038e2dbf1d7342b2632719fc0e424840c323c41

https://github.com/PHPUnitGoodPractices/polyfill/pull/17/checks?check_run_id=1422129120#step:7:4

Reply:
> I did initially like you proposing, but it doesn't work. Just look at the code!

please, calm and easy

> Do I have to copy the code here? Here we go:

no, link is more than enough

> Note that `expectExceptionMessageRegExp` does not set `$this->expectedException`.

**Exactly**, and here, we are polyfilling `expectExceptionMessageMatches`, which does not set exception as well, only the message https://github.com/sebastianbergmann/phpunit/blob/eafcb3a19339f99232354dbf949bd2f890f2491e/src/Framework/TestCase.php#L619


If you want to call `expectExceptionMessageMatches` or `expectExceptionMessageRegExp`, you should expect exception class in first place, and as you are not doing it in the test, the test is failing. take a look at example at source: https://github.com/sebastianbergmann/phpunit/blob/e15dfe6ebddacd7ed093b008332f775a4cc077f3/tests/unit/Framework/Constraint/ExceptionMessageRegExpTest.php#L21-L27

Reply:
Thank you for the correction. Let me research and correct the tests.

Reply:
Right. The thing is `expectExceptionMessageMatches` catches an exception even if you haven't defined specific type.

Reply:
Therefore if you write test like so in the current version of PHPUnit:

```php
    public function testExpectExceptionMessageMatches()
    {
        $this->expectExceptionMessageMatches('/example/');

        throw new \RuntimeException('example');
    }
```
It works as expected without ` $this->expectException(\RuntimeException::class);`

Reply:
If you want the same test to work correctly in PHPUnit 5.7.27, there has to be something more than just calling `expectExceptionMessageMatches`.

Reply:
In latest PHPUnit there's [a logic that considers all possible cases](https://github.com/sebastianbergmann/phpunit/blob/eafcb3a19339f99232354dbf949bd2f890f2491e/src/Framework/TestCase.php#L2487-L2489) for exception expectations. 

Where earlier[ it required an exception type to be present](https://github.com/sebastianbergmann/phpunit/blob/b7803aeca3ccb99ad0a506fa80b64cd6a56bbc0c/src/Framework/TestCase.php#L646) to consider its message.

Reply:
easy solution, maybe good enough for polyfill - the method should set the message only, and if it has to provide the exception class too, it sets `\Exception`.

better solution, but ignore it if too tricky to implement - we set exception message only, and if it has to provide the exception class too, it check what exception class is currently checked and it re-uses it.